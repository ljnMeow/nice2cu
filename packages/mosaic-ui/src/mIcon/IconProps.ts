import { ExtractPropTypes, PropType } from 'vue';

export const IconProps = {
	icon: {
		type: String,
		required: true,
	},
	classPrefix: {
		type: String,
		default: () => 'mosaic-icon',
	},
	margin: [String, Number] as PropType<string | number>,
	badge: [Boolean, Number, String] as PropType<boolean | number | string>,
	multicolor: Boolean,
	color: String,
	size: [String, Number] as PropType<string | number>,
} as const;

export type IconProps = ExtractPropTypes<typeof IconProps>;
