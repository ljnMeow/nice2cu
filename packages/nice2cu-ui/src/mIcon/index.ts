import _icon from './Icon.vue';
import { withInstall } from '../../utils/withInstall';

const mIcon = withInstall(_icon);

export default mIcon;

export { mIcon };

declare module 'vue' {
	export interface GlobalComponents {
		mIcon: typeof mIcon;
	}
}
