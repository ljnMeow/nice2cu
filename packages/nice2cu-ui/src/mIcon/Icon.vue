<template>
	<i v-if="propsData.multicolor" :class="bem.b()" :style="{ margin: style.margin }">
		<span class="svg-content">
			<svg :class="bem.m('svg')" :style="{ ...style, margin: 0 }">
				<use :xlink:href="`#${propsData.icon}`"></use>
			</svg>
			<div v-if="propsData.badge" :class="bem.m('badge')">{{ typeof propsData.badge === 'boolean' ? '' : propsData.badge }}</div>
		</span>
		<slot></slot>
	</i>
	<i v-else :class="className" :style="style">
		<slot></slot>
		<div v-if="propsData.badge" :class="bem.m('badge')">{{ typeof propsData.badge === 'boolean' ? '' : propsData.badge }}</div>
	</i>
</template>

<script lang="ts">
import { defineComponent, computed, ComputedRef } from 'vue';
import { createNamespace } from '../../utils/create';
import { handleUnit } from '../../utils/tools';
import { IconProps } from './IconProps';
import './style/icon.less';
import './assets/font';

export default defineComponent({
	name: 'MIcon',
	props: IconProps,
	setup(props: IconProps) {
		const bem = createNamespace('icon');
		const propsData = { ...props };
		const className: ComputedRef<string[]> = computed(() => {
			return [bem.b(), props.classPrefix, props.icon];
		});
		const style = computed(() => {
			if (props.multicolor) {
				return {
					...(props.margin ? { margin: handleUnit(props.margin) } : {}),
					...(props.size ? { width: handleUnit(props.size) } : {}),
					...(props.size ? { height: handleUnit(props.size) } : {}),
				};
			} else {
				return {
					...(props.size ? { 'font-size': handleUnit(props.size) } : {}),
					...(props.margin ? { margin: handleUnit(props.margin) } : {}),
					...(props.color ? { color: props.color } : {}),
				};
			}
		});
		return { propsData, bem, className, style };
	},
});
</script>
