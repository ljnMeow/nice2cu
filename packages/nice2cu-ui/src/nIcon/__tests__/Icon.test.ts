import { shallowMount } from '@vue/test-utils';
import { describe, expect, test, vi } from 'vitest';
import { createApp } from 'vue';

document.body.innerHTML = '<script></script>';

import mIcon from '../index';

test('test icon plugin', () => {
	const app = createApp({}).use(mIcon);
	expect(app.component(mIcon.name)).toBeTruthy();
});

test('test icon onClick', () => {
	const onClick = vi.fn();
	const wrapper = shallowMount(mIcon, {
		props: {
			icon: 'n-alarm-sharp',
			onClick,
		},
	});
	wrapper.trigger('click');
	expect(onClick).toHaveBeenCalledTimes(1);
	wrapper.unmount();
});

describe('test Icon Props', () => {
	test('test Icon icon', () => {
		const wrapper = shallowMount(mIcon, {
			props: {
				icon: 'n-alarm-sharp',
			},
		});
		expect(wrapper.find('.n-icon').classes()).toContain('n-alarm-sharp');
		wrapper.unmount();
	});

	test('test Icon Size', async () => {
		const wrapper = shallowMount(mIcon, {
			props: {
				icon: 'n-alarm-sharp',
				size: 16,
			},
		});
		expect(wrapper.find('.n-icon').attributes('style')).toContain('font-size: 16px;');
		await wrapper.setProps({ icon: 'n-alarm-sharp', size: '20px' });
		expect(wrapper.find('.n-icon').attributes('style')).toContain('font-size: 20px;');
		await wrapper.setProps({ icon: 'n-alarm-sharp', size: '16rem' });
		expect(wrapper.find('.n-icon').attributes('style')).toContain('font-size: 16rem;');
		await wrapper.setProps({ icon: 'n-alarm-sharp', size: '16em' });
		expect(wrapper.find('.n-icon').attributes('style')).toContain('font-size: 16em;');
		await wrapper.setProps({ icon: 'n-alarm-sharp', size: '16vw' });
		expect(wrapper.find('.n-icon').attributes('style')).toContain('font-size: 16vw;');
		wrapper.unmount();
	});

	test('test Icon Color', () => {
		const wrapper = shallowMount(mIcon, {
			props: {
				icon: 'n-alarm-sharp',
				color: '#e2e2e2',
			},
		});
		expect(wrapper.find('.n-icon').attributes('style')).toContain('color: #e2e2e2;');
		wrapper.unmount();
	});

	test('test Icon Margin', () => {
		const wrapper = shallowMount(mIcon, {
			props: {
				icon: 'n-alarm-sharp',
				margin: 10,
			},
		});
		expect(wrapper.find('.n-icon').attributes('style')).toContain('margin: 10px;');
		wrapper.unmount();
	});

	test('test Icon badge', () => {
		const wrapper = shallowMount(mIcon, {
			props: {
				icon: 'n-alarm-sharp',
				badge: true,
			},
		});
		expect(wrapper.find('.n-icon--badge').html()).toMatchSnapshot();
		wrapper.unmount();
	});

	test('test Icon badge String', () => {
		const wrapper = shallowMount(mIcon, {
			props: {
				icon: 'n-alarm-sharp',
				badge: '10',
			},
		});
		expect(wrapper.find('.n-icon--badge').html()).toMatchSnapshot();
		wrapper.unmount();
	});

	test('test Icon badge Number', () => {
		const wrapper = shallowMount(mIcon, {
			props: {
				icon: 'n-alarm-sharp',
				badge: 10,
			},
		});
		expect(wrapper.find('.n-icon--badge').html()).toMatchSnapshot();
		wrapper.unmount();
	});

	test('test Icon multicolor', () => {
		const wrapper = shallowMount(mIcon, {
			props: {
				icon: 'n-alarm-sharp',
				multicolor: true,
			},
		});
		expect(wrapper.find('.n-icon').html()).toMatchSnapshot();
		wrapper.unmount();
	});

	test('test Icon ClassPrefix', () => {
		const wrapper = shallowMount(mIcon, {
			props: {
				icon: 'n-alarm-sharp',
				classPrefix: 'xxx-icon',
			},
		});
		expect(wrapper.find('.n-icon').classes()).toContain('xxx-icon');
		wrapper.unmount();
	});
});
