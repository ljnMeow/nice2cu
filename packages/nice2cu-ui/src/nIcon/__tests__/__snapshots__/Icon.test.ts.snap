// Vitest Snapshot v1

exports[`test Icon Props > test Icon badge 1`] = `"<div class=\\"n-icon__badge\\"></div>"`;

exports[`test Icon Props > test Icon badge Number 1`] = `"<div class=\\"n-icon__badge\\">10</div>"`;

exports[`test Icon Props > test Icon badge String 1`] = `"<div class=\\"n-icon__badge\\">10</div>"`;

exports[`test Icon Props > test Icon svg 1`] = `"<i class=\\"n-icon\\"><span class=\\"svg-content\\"><svg class=\\"n-icon__svg\\"><use xlink:href=\\"#n-alarm-sharp\\"></use></svg><!--v-if--></span></i>"`;
